packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -519,7 +562,53 @@ class RenderWrap extends RenderBox
     int childCount = 0;
     RenderBox? child = firstChild;
     while (child != null) {
-      final Size childSize = layoutChild(child, childConstraints);
+      Size childSize;
+
+      final WrapParentData childParentData = child.parentData! as WrapParentData;
+      WrapFit fit = childParentData.fit;
+
+      if (mainAxisLimit.isInfinite) {
+        switch (fit) {
+          case WrapFit.runTight:
+            fit = WrapFit.runLoose;
+          case WrapFit.tight:
+            fit = WrapFit.loose;
+          case WrapFit.loose:
+          case WrapFit.runLoose:
+        }
+      }
+
+      switch (fit) {
+        case WrapFit.runTight:
+        case WrapFit.runLoose:
+          final double childIntrinsicMainAxisExtent = getChildMinIntrinsicMainAxisExtent(child, double.infinity);
+          if (runMainAxisExtent != 0 && runMainAxisExtent + childIntrinsicMainAxisExtent + spacing <= mainAxisLimit) {
+            switch (fit) {
+              case WrapFit.runTight:
+                childSize = layoutChild(child, childConstraintsFittingTightInRun(mainAxisLimit - runMainAxisExtent - spacing));
+              case WrapFit.runLoose:
+                childSize = layoutChild(child, childConstraintsFittingLooseInRun(mainAxisLimit - runMainAxisExtent - spacing));
+              case WrapFit.tight:
+              case WrapFit.loose:
+                throw Exception('Unreachable!');
+            }
```

Comment:
I find the nested switches make this code a little harder to follow. I wonder if we can refactor this.

Reply:
Is this better? a1edd737f35f61b4ccd9bd048c7a516ddcfade8f